larpandoracontent/LArTwoDReco/LArCosmicRay/TrackExtensionRefinementAlgorithm.cc

```diff
@@ -0,0 +1,416 @@
+/**
+ *  @file   larpandoracontent/LArTwoDReco/LArCosmicRay/TrackExtensionRefinementAlgorithm.cc
+ *
+ *  @brief  Implementation of the track refinement class.
+ *
+ *  $Log: $
+ */
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArTwoDReco/LArCosmicRay/TrackExtensionRefinementAlgorithm.h"
+
+#include "larpandoracontent/LArControlFlow/MultiPandoraApi.h"
+
+#include "larpandoracontent/LArHelpers/LArGeometryHelper.h"
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+#include "larpandoracontent/LArHelpers/LArHitWidthHelper.h"
+#include "larpandoracontent/LArHelpers/LArStitchingHelper.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+TrackExtensionRefinementAlgorithm::TrackExtensionRefinementAlgorithm() :
+    m_maxLoopIterations(10),    
+    m_growingFitInitialLength(10.f),
+    m_growingFitSegmentLength(5.0f),
+    m_distanceToLine(0.5f),
+    m_boundaryTolerance(2.f)   
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------    
+    
+StatusCode TrackExtensionRefinementAlgorithm::Run()
+{
+    const ClusterList *pClusterList(nullptr);
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*this, pClusterList));
+
+    const CaloHitList *pCaloHitList(nullptr);
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*this, pCaloHitList));
+
+    ClusterVector clusterVector;
+    TwoDSlidingFitResultMap microSlidingFitResultMap, macroSlidingFitResultMap;
+    SlidingFitResultMapPair slidingFitResultMapPair({&microSlidingFitResultMap, &macroSlidingFitResultMap});
+
+    this->InitialiseGeometry();
+    
+    this->InitialiseContainers(pClusterList, SortByDistanceToTPCBoundary(m_tpcMinXEdge), clusterVector, slidingFitResultMapPair);
+
+    // ATTN: Keep track of created main track clusters so their hits can be protected in future iterations
+    ClusterList createdMainTrackClusters;
+    for (unsigned int isHigherXBoundary = 0; isHigherXBoundary < 2; ++isHigherXBoundary)
```

Comment:
You can replace this by a loop of the form:

`for (bool isHigherXBoundary : { false, true }) `

which is the logic I think you're aiming for?

Reply:
indeed, thank you!